### DISCUSSION TITLE
Copy `tfsec` exact behavior with `trivy config`

### DISCUSSION THREAD

**Discussion by hugodocto** [aquasecurity/trivy#5060] [Q&A] [answered]
### Question

Hi there,

following the announcement saying that [tfsec is joining the trivy family](https://github.com/aquasecurity/tfsec/discussions/1994), we're planning to migrate and replace all our `tfsec` homemade Github actions by `trivy` (also homemade).

However, even when runnning `trivy` with similar configuration, we can't make the output the same. The following command using `tfsec` tells that no problem has been detected, and is passing our CI checks:

```
tfsec --config-file tfsec_exclude_list.yml --exclude-downloaded-modules .
``` 

The similar check using `trivy` looks like:

```
trivy config --tf-exclude-downloaded-modules  --skip-files "./**/*.y*ml" --skip-files "./**/*.tgz"
```

However, this check returns a lot of identified vulnerabilities. So my question is why are the results not the same for the same commands? Am I missing something? Sadly, this is blocking us from migrating from `tfsec` to `trivy` seamlessly as it breaks many CI checks.

**More infos:**

* Both `trivy` and `tfsec` are scanning the same number of blocks, so I guess they are checking exactly the same files. I'm 100% sure `trivy` is scanning only Terraform files.
* The content of `.trivyignore` is the same than the content of `tfsec_exclude_list.yml`.
* We are using some inline filtering, but even when I `sed` them form `tfsec:ignore` (which AFAIK is still supported) to `trivy:ignore`, the issue remains the same.

Thanks!

### Target

Filesystem

### Scanner

Misconfiguration

### Output Format

None

### Mode

Standalone

### Version

```bash
$ trivy --version
Version: 0.44.1
Policy Bundle:
  Digest: sha256:2e95a2d5d45de8ebecae53a97403230a6c608a579b082f3de170f3cf09e46243
  DownloadedAt: 2023-08-28 07:25:18.581938134 +0000 UTC
$ tfsec --version
v1.28.1
```

**Comment by AnaisUrlichs** [comment]
@simar7 could you look into this?

**Comment by simar7** [comment]
hi @hugodocto - when you say Trivy identifies more things than tfsec, are these new results meaningful? We've added many more policies in Trivy than tfsec which could be the reason why you are seeing this behavior. 

Also in Trivy we've improved on bugs that existed in tfsec's scanning, which results in proper evaluation of your IaC. It could very well be that tfsec was simply not scanning everything correctly you had and Trivy does. But without seeing the scan results it's hard to say.

Happy to help if you could provide some more details as to what you are seeing.

cc @nikpivkin

**Comment by hugodocto** [comment]
hi @simar7! 

>  when you say Trivy identifies more things than tfsec, are these new results meaningful?

Yes they are. It's misconfiguration issues, and severity levels go from low to high, so probably things we want to handle and fix in the near future. However, due to the number of TF files that are spread around multiple repos, doing the change will break a lot of things a and prevent any new merge for a significant amount of time and effort.

For more context, here are the (sanitized) outputs of both `tfsec` and `trivy` launched with the commands above:

* `tfsec`:

```
  ──────────────────────────────────────────
  disk i/o             ***
  parsing              ***
  adaptation           ***
  checks               ***
  total                ***

  counts
  ──────────────────────────────────────────
  modules downloaded   0
  modules processed    ***
  blocks processed     ***
  files read           ***

  results
  ──────────────────────────────────────────
  passed               ***
  ignored              ***
  critical             0
  high                 0
  medium               0
  low                  0


No problems detected!
```

* `trivy`:

```
2023-09-01T16:16:33.050+0200	INFO	Misconfiguration scanning is enabled
2023-09-01T16:23:31.452+0200	INFO	Detected config files: ***

[found issues listing]
```

I did the calculation and `trivy` now finds 42 vulnerabilities where `tfsec` founds zero.

But if you added policies for Trivy, this totally makes sense. Maybe there is a way to disable those policies and re-enables them in the future?

**Reply by nikpivkin** [reply]
Hi @hugodocto !

You can filter the results using [ignore files](https://aquasecurity.github.io/trivy/v0.45/docs/configuration/filtering/#by-finding-ids) or [inline comments](https://aquasecurity.github.io/trivy/v0.45/docs/configuration/filtering/#by-inline-comments).

**Reply by hugodocto** [reply]
Hi @nikpivkin
indeed, adding more inline comments or entries in the `.trivyignore` file may do the job. However, this will require a lot of tweaking.
We have a common `tfsec_exclude_list.yml` that serves as a base for all repos, and then inline comments locally for each block that needs a special exception.

The example I pasted above is only one of our many repositories containing terraform code, so I guess we we'll have a lot more new vulns depending on the content of the repos.

But we could _in theory_ build a big `.trivyignore` allowlist indeed

**Reply by simar7** [reply, marked as answer]
We've added some improvements to `.trivyignore` in the latest release. Please have a look https://aquasecurity.github.io/trivy/v0.45/docs/configuration/filtering/#trivyignoreyaml
